src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
 
 /*
+ * Copyright (C) Evgenii Sokolov
  * Copyright (C) Igor Sysoev
  * Copyright (C) NGINX, Inc.
  */
 
 #include <nxt_main.h>
 
 
+static void nxt_array_elts_copy(void **dst_elts, nxt_array_t *src_array);
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
 nxt_array_t *
-nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
+nxt_array_create(nxt_mp_t *mp, nxt_uint_t arr_size, size_t elt_size)
 {
-    nxt_array_t  *array;
+    nxt_uint_t     nalloc;
+    nxt_array_t    *array;
+    void           *elts;
 
-    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);
+    array = nxt_mp_alloc(mp, sizeof(nxt_array_t));
 
     if (nxt_slow_path(array == NULL)) {
         return NULL;
     }
 
-    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
+    nalloc = nxt_array_calc_alloc(arr_size);
+    elts = nxt_mp_alloc(mp, elt_size * nalloc);
```

Comment:
- fixed initialization of a set of elements (elts), pointer, memory size;

Reply:
Logically, I will think about how to improve this in the context of the changes I propose.

But I want to note that in the changes I propose, the allocated and unused memory at the time of creating the array is released in the event of its subsequent expansion, and in the context of existing code, this memory area is not released at all and is not used after it is allocated from the moment the array is created, regardless of the need to change the value.

Reply:
It makes only sense to release big amounts of memory, at least more than a page size. All small allocations will be released when the memory pool be destroyed. Those small allocations don't make any visible difference to the memory consumption during request processing, but any additional calls can introduce additional delays.

Reply:
Apart from what @VBart said, allocating more than what you need is actually a good thing, not only because of the direct impact of the reduced number of `malloc(3)` and `free(3)` calls, but because of how it works underneath:

The system tries to find a contiguous block of memory that is big enough for what you asked.  If you ask big chunks of memory (especially good if you allocate 4K blocks), it's especially easy for the system.  Then, you deallocate big chunks at once, and the free memory tables receive the full blocks of memory at once, and can reuse them easily.  If you allocate small amounts of memory, and you deallocate them as soon as you can, then the free memory blocks become small and disperse, which becomes a problem because the allocator has to check each of those small chunks to see if they meet the requirements.  That's the same thing that happen(ed) to hard drives: fragmentation.

Yes, we do explicitly waste memory, but systems have more and more memory these days.  I guess we can waste some.  The benefits are:  less calls to `malloc(3)` and `free(3)` *and* each of those few calls is faster because of having "defragmented" memory.

Reply:
Now I understand. Thank you for the detailed explanation. I thought about it, but I did not understand the principles of defragmentation in Unit so well and the high cost of each individual call. Now it's all clear. Thanks again!